docs/design/2022-06-22-foreign-key.md

```diff
@@ -0,0 +1,778 @@
+);
+
+INSERT INTO customers_2 VALUES (1), (2), (3);
+INSERT INTO orders_2 VALUES (100,1), (101,2), (102,3), (103,1);
+```
+
+```sql
+> explain analyze UPDATE customers_2 SET id = 23 WHERE id = 1;
+                       info
+--------------------------------------------------
+  planning time: 494µs
+  execution time: 5ms
+  distribution: local
+  vectorized: true
+  rows read from KV: 6 (170 B)
+  cumulative time spent in KV: 978µs
+  maximum memory usage: 100 KiB
+  network usage: 0 B (0 messages)
+  regions: us-east1
+
+  • root
+  │
+  ├── • update
+  │   │ nodes: n1
+  │   │ regions: us-east1
+  │   │ actual row count: 1
+  │   │ table: customers_2
+  │   │ set: id
+  │   │
+  │   └── • buffer
+  │       │ label: buffer 1
+  │       │
+  │       └── • render
+  │           │ nodes: n1
+  │           │ regions: us-east1
+  │           │ actual row count: 1
+  │           │ KV rows read: 1
+  │           │ KV bytes read: 27 B
+  │           │
+  │           └── • scan
+  │                 nodes: n1
+  │                 regions: us-east1
+  │                 actual row count: 1
+  │                 KV rows read: 1
+  │                 KV bytes read: 27 B
+  │                 missing stats
+  │                 table: customers_2@primary
+  │                 spans: [/1 - /1]
+  │                 locking strength: for update
+  │
+  └── • fk-cascade
+        fk: fk_customer_id_ref_customers_2
+        input: buffer 1
+```
+
+##### PostgreSQL DML Execution Plan
+
+```sql
+postgres=# explain analyze UPDATE customers_2 SET id = 20 WHERE id = 23;
+                                                             QUERY PLAN
+-------------------------------------------------------------------------------------------------------------------------------------
+ Update on customers_2  (cost=0.15..8.17 rows=1 width=10) (actual time=0.039..0.039 rows=0 loops=1)
+   ->  Index Scan using customers_2_pkey on customers_2  (cost=0.15..8.17 rows=1 width=10) (actual time=0.016..0.016 rows=1 loops=1)
+         Index Cond: (id = 23)
+ Planning Time: 0.057 ms
+ Trigger for constraint orders_2_customer_id_fkey on customers_2: time=0.045 calls=1
+ Trigger for constraint orders_2_customer_id_fkey on orders_2: time=0.023 calls=2
+ Execution Time: 0.129 ms
+```
+
+## Other Technical Design
+
+### How to check foreign key integrity?
+
+How MySQL to do this? Look like MySQL doesn't provide any method, but the user can use stored procedure to do this, see: https://stackoverflow.com/questions/2250775/force-innodb-to-recheck-foreign-keys-on-a-table-tables
+
+Maybe We can use following syntax to check foreign key integrity:
+
+```sql
+ADMIN CHECK FOREIGN KEY [table_name] [foreign_key_name]
+```
+
+which implemention is use following SQL to check:
+
+```sql
+select count(*) from t where t.a not in (select id from t_refer);
+```
+
+## Impact
+
+### Impact of data replication
+
+todo
+
```

Comment:
For one-time data load/import/restore, if the implementation is based on SQL layer (for example, lightning tidb backend), the behaviour should be the same as the behavour in "DML Load data" section. If it's based on storage later (for example, lightning local backend), we do not guarantee any foreign key properties, including:
- inserted rows has no respect for foreign key, as `set @@foreign_key_checks=0`
- overwritten rows will also not trigger referenced table actions as in section "DML On reference Table"

Reply:
For long-time running data replication like CDC/DM, I can list some replication standards in the order of enabling more functionality:

1. mask out FK features when replicating
2. replicate FK related statements as-is, but in downstream runtime disable FK features
3. the FK features is correctly enabled during replicating

Seems currently standard level 2 is easy to accomplish, and the shortcoming is DML on referenced table will not trigger ["Referential Actions"](https://dev.mysql.com/doc/refman/8.0/en/create-table-foreign-keys.html#foreign-key-referential-actions) like DELETE CASCADE or SET NULL. Is it acceptable? @frank945946

And please help us check if these criterias can meet @crazycs520 
- There's a way to execute DDL with FK just like it does not contain FK. Specifically, can drop the index required by a foreign key
- There's a way to do DML on child tables like it does not contain FK
- There's a way to do DML on referenced table like it does not contain FK

Reply:
>Seems currently standard level 2 is easy to accomplish, and the shortcoming is DML on referenced table will not trigger ["Referential Actions"](https://dev.mysql.com/doc/refman/8.0/en/create-table-foreign-keys.html#foreign-key-referential-actions) like DELETE CASCADE or SET NULL. Is it acceptable?

For data replication like CDC, since CDC watch the kv change then generage SQL to to data replication in downstream databases, so disable FK feature in downstream is ok, since if upstream enable FK feature and does Referential Actions, CDC will generate addiational SQL about it.

For data replication like DM, I think we should confirm the binlog of DML with FK, 
* if the binlog contains the FK Referential Actions, then disable FK feature in downstream is ok and generate additional SQL executed in downstream.
*  If the binlog doesn't contain the FK Referential Actions, then downstream maybe should enable FK feature.

Reply:
>And please help us check if these criterias can meet @crazycs520
>- There's a way to execute DDL with FK just like it does not contain FK. Specifically, can drop the index required by a foreign key
>- There's a way to do DML on child tables like it does not contain FK
>- There's a way to do DML on referenced table like it does not contain FK

Mostly, after setting `set @@foreign_key_checks=0`, mostly DDL and all DML will action like it doesn't contain FK.

A special case is `DROP INDEX` DDL, if the index is used by FK, then you can't drop the index no matter how.